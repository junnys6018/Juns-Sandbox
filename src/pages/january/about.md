This is an implementation of an old rendering technique called voxel space rendering commonly used in the 1990's

### The math

<p class="flex flex-row justify-center gap-8">
<img src="/media/january/Aztec-Color.png" alt="color map" class="w-1/3 inline-block">
<img src="/media/january/Aztec-Depth.png" alt="depth map" class="w-1/3 inline-block">
</p>

We are given a height and color map, call it $c(p)$ and $d(p)$. Where $p\in[0,1]\times[0,1]$ is the point on the map to be sampled. Suppose the camera is at some point $c\in[0,1]\times[0,1]$, with height $h$. Voxel space rendering works by scanning the view frustum in a back to front order, for each point $p$ on the frustum, we calculate the height of the rasterized point with the formula,

$$
\frac{d(p)-h}{z}\qquad\qquad z=||c-p||
$$

Where $d(p)-h$ is the height of the point relative to the camera, and $z$ is a scaling term used to perform perspective projection. We can adjust the pitch of the camera by adding an offset term, $o$ to the formula,

$$
\frac{d(p)-h}{z}+o\qquad\qquad z=||c-p||
$$

Effectively translating the image by a fixed value. We draw a vertical line from the bottom of the rasterized image to the calculated height with the sampled color $c(p)$.

### References

[https://github.com/s-macke/VoxelSpace](https://github.com/s-macke/VoxelSpace)
