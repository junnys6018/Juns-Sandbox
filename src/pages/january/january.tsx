import Footer from 'components/footer';
import Heading from 'components/heading';
import Navbar from 'components/navbar';
import Socials from 'components/socials';
import Slider from 'components/slider';

import Aztec from './maps/aztec.png';
import Canyon from './maps/canyon.png';
import Plains from './maps/plains.png';
import Tundra from './maps/tundra.png';
import voxelSpaceApi from './wasm/voxel-space-api';
import aboutMarkdown from './about.md';
import { AsyncMarkdown } from 'markdown';

import { RadioGroup } from '@headlessui/react';
import { Fragment, useEffect, useRef, useState } from 'react';
import classNames from 'classnames';

function MapOption(props: { value: string; src: string }) {
    return (
        <RadioGroup.Option
            value={props.value}
            className="group flex flex-col items-center focus-visible:outline-none cursor-pointer"
        >
            {({ checked }) => (
                <Fragment>
                    <img
                        alt={props.value}
                        src={props.src}
                        className={classNames(
                            'rounded-lg transition',
                            !checked && 'brightness-75 group-hover:brightness-100',
                        )}
                    ></img>
                    <RadioGroup.Label
                        className={classNames(
                            'font-medium text-sm cursor-pointer',
                            checked ? 'text-pink-500' : 'text-neutral-700',
                        )}
                    >
                        {props.value}
                    </RadioGroup.Label>
                </Fragment>
            )}
        </RadioGroup.Option>
    );
}

function PathOption(props: { value: string; className?: string }) {
    return (
        <RadioGroup.Option
            value={props.value}
            className={({ checked }) =>
                classNames(
                    'cursor-pointer text-base sm:text-xl py-1.5 px-4 focus-visible:outline-none',
                    checked ? 'bg-pink-500 text-neutral-50 drop-shadow-pink-sm' : 'bg-neutral-200 text-neutral-800',
                    props.className,
                )
            }
        >
            <RadioGroup.Label className="cursor-pointer">{props.value}</RadioGroup.Label>
        </RadioGroup.Option>
    );
}

type Map = 'Aztec' | 'Tundra' | 'Plains' | 'Canyon';
type Path = 'Rotate' | 'Linear' | 'Circle';

export interface SettingsPanelProps {
    map: Map;
    path: Path;
    pitch: number;
    cameraHeight: number;

    setMap: React.Dispatch<React.SetStateAction<Map>>;
    setPath: React.Dispatch<React.SetStateAction<Path>>;
    setPitch: React.Dispatch<React.SetStateAction<number>>;
    setCameraHeight: React.Dispatch<React.SetStateAction<number>>;
}

export function SettingsPanel(props: SettingsPanelProps) {
    return (
        <div className="drop-shadow-lg bg-white rounded-3xl p-8 flex-shrink-0 w-full xl:w-[470px]">
            <h3 className="font-medium text-neutral-900 mb-2.5">Map</h3>
            <RadioGroup value={props.map} onChange={props.setMap} className="grid grid-cols-4 gap-2 sm:gap-4 mb-4">
                <MapOption value="Aztec" src={Aztec} />
                <MapOption value="Tundra" src={Tundra} />
                <MapOption value="Plains" src={Plains} />
                <MapOption value="Canyon" src={Canyon} />
            </RadioGroup>
            <h3 className="font-medium text-neutral-900 mb-2.5">Path</h3>
            <RadioGroup value={props.path} onChange={props.setPath} className="inline-flex flex-row mb-4">
                <PathOption value="Rotate" className="rounded-l-lg" />
                <PathOption value="Linear" />
                <PathOption value="Circle" className="rounded-r-lg" />
            </RadioGroup>
            <h3 className="font-medium text-neutral-900 mb-2.5">Pitch</h3>
            <Slider
                min={0.2}
                max={1.5}
                step={0.02}
                value={props.pitch}
                onChange={e => props.setPitch(parseFloat(e.currentTarget.value))}
                className="w-full mb-4"
            ></Slider>
            <h3 className="font-medium text-neutral-900 mb-2.5">Camera height</h3>
            <Slider
                min={30}
                max={150}
                step={1}
                value={props.cameraHeight}
                onChange={e => props.setCameraHeight(parseFloat(e.currentTarget.value))}
                className="w-full"
            ></Slider>
        </div>
    );
}

type Animate = {
    [key in Path]: (timestamp: number) => [number, number, number];
};

const animate: Animate = {
    Rotate: (timestamp: number) => [timestamp / 8000, 170, 700],
    Linear: (timestamp: number) => [0, timestamp / 10, 700],
    Circle: (timestamp: number) => {
        const RADIUS = 600;
        const PI_OVER_2 = 1.57079632679;
        const t = timestamp / 8000;
        const xpos = RADIUS * Math.cos(t);
        const ypos = RADIUS * Math.sin(t);
        const phi = t + PI_OVER_2;
        return [phi, xpos, ypos];
    },
};

const canvasWidth = 600;
const canvasHeight = 400;

export default function January() {
    const [map, setMap] = useState<Map>('Aztec');
    const [path, setPath] = useState<Path>('Rotate');
    const [pitch, setPitch] = useState(0.7);
    const [cameraHeight, setCameraHeight] = useState(100);

    const canvasElement = useRef<HTMLCanvasElement>(null);
    const voxelSpaceContext = useRef<number | null>(null);

    useEffect(() => {
        voxelSpaceApi.then(([, api]) => {
            voxelSpaceContext.current = api.createContext(`maps/${map}-Color.png`, `maps/${map}-Depth.png`, 0xffe09090);
        });

        return () => {
            voxelSpaceApi.then(([, api]) => {
                if (voxelSpaceContext.current !== null) {
                    api.destroyContext(voxelSpaceContext.current);
                }
            });
        };
    }, [map]);

    useEffect(() => {
        let rafId: number;

        voxelSpaceApi.then(([instance, api]) => {
            const canvasContext = canvasElement.current?.getContext('2d');

            const raf = (timestamp: number) => {
                rafId = requestAnimationFrame(raf);

                if (voxelSpaceContext.current === null || canvasContext === undefined || canvasContext === null) {
                    return;
                }

                const image = api.render(
                    voxelSpaceContext.current,
                    canvasWidth,
                    canvasHeight,
                    ...animate[path](timestamp),
                    pitch,
                    cameraHeight,
                );

                const imageView = new Uint8ClampedArray(instance.HEAPU8.buffer, image, canvasWidth * canvasHeight * 4);
                const imageData = new ImageData(imageView, canvasWidth, canvasHeight);

                canvasContext.putImageData(imageData, 0, 0);
            };

            rafId = requestAnimationFrame(raf);
        });

        return () => {
            cancelAnimationFrame(rafId);
        };
    }, [cameraHeight, pitch, path]);

    return (
        <div className="flex flex-col min-h-screen">
            <Navbar />
            <div className="container mb-8">
                <Heading month={1} />
                <div className="flex flex-row flex-wrap mt-6 gap-x-10 gap-y-6">
                    <canvas
                        className="flex-grow pixelated max-w-full"
                        width={canvasWidth}
                        height={canvasHeight}
                        ref={canvasElement}
                    ></canvas>
                    <SettingsPanel
                        map={map}
                        path={path}
                        pitch={pitch}
                        cameraHeight={cameraHeight}
                        setMap={setMap}
                        setPath={setPath}
                        setPitch={setPitch}
                        setCameraHeight={setCameraHeight}
                    />
                </div>
                <h1 className="font-semibold text-4xl text-pink-500 my-8">About this demo</h1>
                <AsyncMarkdown className="max-w-none" src={aboutMarkdown} />
            </div>
            <Footer />
            <Socials />
        </div>
    );
}
