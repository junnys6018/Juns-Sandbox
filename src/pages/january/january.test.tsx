import { render, screen } from '@testing-library/react';
import { SettingsPanel, SettingsPanelProps } from './january';
import userEvent from '@testing-library/user-event';

test('<SettingsPanel /> snapshot', () => {
    const props: SettingsPanelProps = {
        map: 'Aztec',
        path: 'Circle',
        pitch: 0,
        cameraHeight: 0,

        setMap: jest.fn(),
        setPath: jest.fn(),
        setPitch: jest.fn(),
        setCameraHeight: jest.fn(),
    };
    const { asFragment } = render(<SettingsPanel {...props} />);
    expect(asFragment()).toMatchSnapshot();
});

test('<SettingsPanel /> user interaction', () => {
    const props: SettingsPanelProps = {
        map: 'Aztec',
        path: 'Circle',
        pitch: 0,
        cameraHeight: 0,

        setMap: jest.fn(),
        setPath: jest.fn(),
        setPitch: jest.fn(),
        setCameraHeight: jest.fn(),
    };

    render(<SettingsPanel {...props} />);

    const plainsLabel = screen.getByLabelText('Plains');
    userEvent.click(plainsLabel);
    expect(props.setMap).toBeCalledWith('Plains');

    const linearPathLabel = screen.getByLabelText('Linear');
    userEvent.click(linearPathLabel);
    expect(props.setPath).toBeCalledWith('Linear');
});
