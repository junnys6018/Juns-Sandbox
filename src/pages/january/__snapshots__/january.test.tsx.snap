// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SettingsPanel /> snapshot 1`] = `
<DocumentFragment>
  <div
    class="mx-auto drop-shadow-lg bg-white rounded-3xl p-8"
    style="width: 470px; margin-top: 100px;"
  >
    <h3
      class="font-medium text-neutral-900 mb-2.5"
    >
      Map
    </h3>
    <div
      class="flex flex-row gap-6 mb-4"
      id="headlessui-radiogroup-9"
      role="radiogroup"
    >
      <div
        aria-checked="true"
        aria-labelledby="headlessui-label-1"
        class="group flex flex-col items-center focus-visible:outline-none cursor-pointer"
        id="headlessui-radiogroup-option-2"
        role="radio"
        tabindex="0"
      >
        <img
          alt="Aztec"
          class="rounded-lg transition"
          src="aztec.png"
        />
        <label
          class="font-medium text-sm cursor-pointer text-pink-500"
          id="headlessui-label-1"
        >
          Aztec
        </label>
      </div>
      <div
        aria-checked="false"
        aria-labelledby="headlessui-label-3"
        class="group flex flex-col items-center focus-visible:outline-none cursor-pointer"
        id="headlessui-radiogroup-option-4"
        role="radio"
        tabindex="-1"
      >
        <img
          alt="Tundra"
          class="rounded-lg transition brightness-75 group-hover:brightness-100"
          src="tundra.png"
        />
        <label
          class="font-medium text-sm cursor-pointer text-neutral-700"
          id="headlessui-label-3"
        >
          Tundra
        </label>
      </div>
      <div
        aria-checked="false"
        aria-labelledby="headlessui-label-5"
        class="group flex flex-col items-center focus-visible:outline-none cursor-pointer"
        id="headlessui-radiogroup-option-6"
        role="radio"
        tabindex="-1"
      >
        <img
          alt="Plains"
          class="rounded-lg transition brightness-75 group-hover:brightness-100"
          src="plains.png"
        />
        <label
          class="font-medium text-sm cursor-pointer text-neutral-700"
          id="headlessui-label-5"
        >
          Plains
        </label>
      </div>
      <div
        aria-checked="false"
        aria-labelledby="headlessui-label-7"
        class="group flex flex-col items-center focus-visible:outline-none cursor-pointer"
        id="headlessui-radiogroup-option-8"
        role="radio"
        tabindex="-1"
      >
        <img
          alt="Canyon"
          class="rounded-lg transition brightness-75 group-hover:brightness-100"
          src="canyon.png"
        />
        <label
          class="font-medium text-sm cursor-pointer text-neutral-700"
          id="headlessui-label-7"
        >
          Canyon
        </label>
      </div>
    </div>
    <h3
      class="font-medium text-neutral-900 mb-2.5"
    >
      Path
    </h3>
    <div
      class="inline-flex flex-row mb-4"
      id="headlessui-radiogroup-16"
      role="radiogroup"
    >
      <div
        aria-checked="false"
        aria-labelledby="headlessui-label-10"
        class="cursor-pointer text-xl py-1.5 px-4 focus-visible:outline-none bg-neutral-200 text-neutral-800 rounded-l-lg"
        id="headlessui-radiogroup-option-11"
        role="radio"
        tabindex="-1"
      >
        <label
          class="cursor-pointer"
          id="headlessui-label-10"
        >
          Rotate
        </label>
      </div>
      <div
        aria-checked="false"
        aria-labelledby="headlessui-label-12"
        class="cursor-pointer text-xl py-1.5 px-4 focus-visible:outline-none bg-neutral-200 text-neutral-800"
        id="headlessui-radiogroup-option-13"
        role="radio"
        tabindex="-1"
      >
        <label
          class="cursor-pointer"
          id="headlessui-label-12"
        >
          Linear
        </label>
      </div>
      <div
        aria-checked="true"
        aria-labelledby="headlessui-label-14"
        class="cursor-pointer text-xl py-1.5 px-4 focus-visible:outline-none bg-pink-500 text-neutral-50 drop-shadow-pink-sm rounded-r-lg"
        id="headlessui-radiogroup-option-15"
        role="radio"
        tabindex="0"
      >
        <label
          class="cursor-pointer"
          id="headlessui-label-14"
        >
          Circle
        </label>
      </div>
    </div>
    <h3
      class="font-medium text-neutral-900 mb-2.5"
    >
      Pitch
    </h3>
    <input
      class="w-full mb-4"
      max="10"
      min="0"
      step="0.01"
      type="range"
      value="0"
    />
    <h3
      class="font-medium text-neutral-900 mb-2.5"
    >
      Camera height
    </h3>
    <input
      class="w-full"
      max="10"
      min="0"
      step="0.01"
      type="range"
      value="0"
    />
  </div>
</DocumentFragment>
`;
